ext/dom/xpath_callbacks.c

```diff
@@ -0,0 +1,504 @@
+		} ZEND_HASH_FOREACH_END();
+	}
+}
+
+static zend_result php_dom_xpath_callback_ns_update_method_handler(php_dom_xpath_callback_ns* ns, xmlXPathContextPtr ctxt, const zend_string *namespace, zend_string *name, const HashTable *callable_ht, php_dom_xpath_callback_name_validation name_validation, php_dom_xpath_callbacks_register_func_ctx register_func)
+{
+	zval *entry, registered_value;
+
+	if (callable_ht) {
+		zend_string *key;
+		ZEND_HASH_FOREACH_STR_KEY_VAL(callable_ht, key, entry) {
+			zend_fcall_info_cache* fcc = emalloc(sizeof(zend_fcall_info));
+			char *error;
+			if (!zend_is_callable_ex(entry, NULL, 0, NULL, fcc, &error)) {
+				zend_argument_type_error(1, "must be an array with valid callbacks as values, %s", error);
+				efree(fcc);
+				efree(error);
+				return FAILURE;
+			}
+
+			zend_fcc_addref(fcc);
+			ZVAL_PTR(&registered_value, fcc);
+
+			if (!key) {
+				zend_string *str = zval_try_get_string(entry);
+				if (str && php_dom_xpath_is_callback_name_valid_and_throw(str, name_validation, true)) {
+					zend_hash_update(&ns->functions, str, &registered_value);
+					if (register_func) {
+						register_func(ctxt, namespace, str);
+					}
+					zend_string_release_ex(str, false);
+				} else {
+					zend_fcc_dtor(fcc);
+					efree(fcc);
+					return FAILURE;
+				}
+			} else {
+				if (!php_dom_xpath_is_callback_name_valid_and_throw(key, name_validation, true)) {
+					zend_fcc_dtor(fcc);
+					efree(fcc);
+					return FAILURE;
+				}
+				zend_hash_update(&ns->functions, key, &registered_value);
+				if (register_func) {
+					register_func(ctxt, namespace, key);
+				}
+			}
+		} ZEND_HASH_FOREACH_END();
+		ns->mode = PHP_DOM_REG_FUNC_MODE_SET;
+	} else if (name) {
+		if (!php_dom_xpath_is_callback_name_valid(name, name_validation)) {
+			zend_argument_value_error(1, "must be a valid callback name");
+			return FAILURE;
+		}
+		zend_fcall_info_cache* fcc = emalloc(sizeof(zend_fcall_info));
+		char *error;
+		zval tmp;
+		ZVAL_STR(&tmp, name);
+		if (!zend_is_callable_ex(&tmp, NULL, 0, NULL, fcc, &error)) {
+			zend_argument_type_error(1, "must be a callable, %s", error);
+			efree(fcc);
+			efree(error);
+			return FAILURE;
+		}
+		zend_fcc_addref(fcc);
+		ZVAL_PTR(&registered_value, fcc);
+		zend_hash_update(&ns->functions, name, &registered_value);
+		if (register_func) {
+			register_func(ctxt, namespace, name);
+		}
+		ns->mode = PHP_DOM_REG_FUNC_MODE_SET;
+	} else {
+		ns->mode = PHP_DOM_REG_FUNC_MODE_ALL;
+	}
+
+	return SUCCESS;
+}
+
+static php_dom_xpath_callback_ns *php_dom_xpath_callbacks_ensure_ns(php_dom_xpath_callbacks *registry, zend_string *ns)
+{
+	if (ns == NULL) {
+		if (!registry->php_ns) {
+			registry->php_ns = emalloc(sizeof(php_dom_xpath_callback_ns));
+			php_dom_xpath_callback_ns_ctor(registry->php_ns);
+		}
+		return registry->php_ns;
+	} else {
+		if (!registry->namespaces) {
+			/* In most cases probably only a single namespace is registered. */
+			registry->namespaces = zend_new_array(1);
```

Comment:
I'm not sure using ``1`` here makes a lot of difference. AFAIK the minimum size if 8? But I might be wrong.

Reply:
Yeah it will up it to 8, but I think I'll keep this because it is an implementation detail of Zend that 1 will be upped to 8.